### PR TITLE
Update to Django 3.2

### PR DESCRIPTION
**Pull request by wenottingham**
##### SUMMARY

AWX currently uses Django 2.2, which goes EOL in April 2022. Being on a security-maintained version of Django is a good thing.

This changeset updates django, several dependencies, applies fixes, and fixes assorted deprecation warnings.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### ADDITIONAL INFORMATION

Here are my notes.

### What changed

Update to django-3.2.
Update to assorted django components that either are required to work with 3.2 or fix deprecation noise.
- asgiref
- django-crum
- django-auth-ldap
- django-oauth-toolkit
- django-jsonfield (see below)
- django-rest-framework-yaml
- django-debug-toolbar (I picked a version with minimal changes. Someone who uses this can pick a different one.)

#### Notes on django-oauth-toolkit

This updates to 1.5, which:
- fixes an issue noted in UPGRADE BLOCKERS
- adds support for OIDC (this could be useful with future SSO work)
- adds new dependencies (wrapt, deprecated, jwcrypto) in support of OIDC and JWT

### Work that still needs done

1. JSONField, part 1:

This note was in UPGRADE BLOCKERS:

```
### django-jsonfield

Instead of calling a `loads()` operation, the returned value is casted into
a string in some cases, introduced in the change:

https://github.com/adamchainz/django-jsonfield/pull/14

This breaks a very large amount of AWX code that assumes these fields
are returned as dicts. Upgrading this library will require a refactor
to accomidate this change.
```

django-jsonfield was a required update due to django changes, and it includes this referenced PR. An attempt was made to resolve this (and it seems to work?), but this should be reviewed.

Note that django-jsonfield is now EOL and in maintenance mode upstream. See next item.

2. JSONField, part 2:

Django 3.1 introduces a new, cross-database, JSONField to replace the postgres-specific version that we inherit from in our JSONBField.

OOTB, we generate warnings like the following in the logs repeatedly:
```
WARNINGS:
main.Credential.inputs: (fields.W904) django.contrib.postgres.fields.JSONField is deprecated. Support for it (except in historical migrations) will be removed in Django 4.0.
	HINT: Use django.db.models.JSONField instead.
main.CredentialInputSource.metadata: (fields.W904) django.contrib.postgres.fields.JSONField is deprecated. Support for it (except in historical migrations) will be removed in Django 4.0.
	HINT: Use django.db.models.JSONField instead.
main.CredentialType.injectors: (fields.W904) django.contrib.postgres.fields.JSONField is deprecated. Support for it (except in historical migrations) will be removed in Django 4.0.
	HINT: Use django.db.models.JSONField instead.
main.CredentialType.inputs: (fields.W904) django.contrib.postgres.fields.JSONField is deprecated. Support for it (except in historical migrations) will be removed in Django 4.0.
	HINT: Use django.db.models.JSONField instead.
main.Host.ansible_facts: (fields.W904) django.contrib.postgres.fields.JSONField is deprecated. Support for it (except in historical migrations) will be removed in Django 4.0.
	HINT: Use django.db.models.JSONField instead.
main.UnifiedJob.installed_collections: (fields.W904) django.contrib.postgres.fields.JSONField is deprecated. Support for it (except in historical migrations) will be removed in Django 4.0.
```

A patch was added to silence the warnings.

We should eventually fix our JSONBField properly (and fix our JSONField as well) to use the new one. If the needed migration is a data migration, it will be painful in terms of migration time.

3. Cleanup

Django 3.1/3.2 make changes under the hood for faster deletion.  Our bulk delete work that inherits from django internals required refactoring to accomodate this. A patch was added that does this, and the unit tests pass and deletion works in the dev env.

This needs audited (I really don't know this code and just made mechanical changes) and performance tested (I may have re-added upstream code that we intentionally left out.)

See https://github.com/django/django/commit/26c66f45193 for the inital upstream change that changed function signatures.

### Unit test failures

```
FAILED awx/main/tests/unit/models/test_credential.py::test_unique_hash_with_unicode - KeyError: 'credential_type'
```

???

### Other notes

The final large commit is just replacing calls to deprecated functions with calls to their replacements. It's a simple search-and-replace, so seemed useful to include. If you want to take it out of this PR and apply it later after other branches merge, go for it.

### CODE REVIEW

File: awx/main/fields.py [hunk 1]
```diff
@@ -78,10 +78,18 @@ def db_type(self, connection):
     def from_db_value(self, value, expression, connection):
         if value in {'', None} and not self.null:
             return {}
-        return super(JSONField, self).from_db_value(value, expression, connection)
+        if isinstance(value, str):
+            return json.loads(value)
+        return value
```
Comment (wenottingham) [inline awx/main/fields.py:83]: For https://github.com/ansible/awx/blob/devel/requirements/README.md#django-oauth-toolkit

File: awx/main/utils/deletion.py [hunk 2]
```diff
@@ -117,26 +123,36 @@ def collect(self, objs, source=None, nullable=False, collect_related=True, sourc
...
+            elif related_qs:
+                field.remote_field.on_delete(self, field, related_qs, self.using)
+        for related_model, related_fields in model_fast_deletes.items():
+            related_qs = self.related_objects(related_model, related_fields, new_objs)
+            self.fast_deletes.append(related_qs)
+        for field in model._meta.private_fields:
+            if hasattr(field, 'bulk_related_objects'):
+                # It's something like generic foreign key.
+                sub_objs = bulk_related_objects(field, new_objs, self.using)
+                self.collect(sub_objs, source=model, nullable=True, fail_on_restricted=False)
```
Comment (wenottingham) [inline awx/main/utils/deletion.py:156]: dont-ask-me

Comment (IPvSean) [reply awx/main/utils/deletion.py:156]: ❤️

**Comment by jbradberry** [comment]
@AlanCoding My inclination is to close and start fresh.  There are merge conflicts, and it's not clear what items could be skipped when going down the list of upgrade notes.

**Comment by AlanCoding** [comment]
yes, sounds good, I just wanted to sync up with you about the status.

**Pull request closed without merging**